File: src/main/java/io/mazenmc/skypebot/stat/StatisticsManager.java
package io.mazenmc.skypebot.stat;

import in.kyle.ezskypeezlife.api.obj.SkypeMessage;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class StatisticsManager {
    private static final StatisticsManager INSTANCE = new StatisticsManager();
    private Map<String, MessageStatistic> statistics = new ConcurrentHashMap<>();

    public static StatisticsManager instance() {
        return INSTANCE;
    }

    public void logMessage(SkypeMessage message) {
        String senderId = message.getSender().getUsername();
        if (!statistics.containsKey(senderId))
            statistics.put(senderId, new MessageStatistic(senderId));

        try {
            statistics.get(senderId).addMessage(message);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public void removeStat(String id) {
        statistics.remove(id);
    }

    public String ownerFor(MessageStatistic statistic) {
        for (Map.Entry<String, MessageStatistic> entry : statistics.entrySet()) {
            if (entry.getValue().equals(statistic))
                return entry.getKey();
        }

        return null;
    }

    public Map<String, MessageStatistic> statistics() {
        return statistics;
    }

    public void loadStatistics() {
        StringBuilder source = new StringBuilder();

        try {
            for (String s : Files.readAllLines(Paths.get("statistics.json"))) {
                source.append(s)
                        .append("\n");
            }
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        try {
            JSONObject object = new JSONObject(source.toString());

            object.keys().forEachRemaining((obj) -> {
                if (!(obj instanceof String)) {
                    return;
                }

                String key = (String) obj;

                try {
                    statistics.put(key, new MessageStatistic(key, object.getJSONArray(key)));
                } catch (JSONException ex) {
                    ex.printStackTrace();
                }
            });
        } catch (JSONException ex) {
            ex.printStackTrace();
        }
    }

    public void saveStatistics() {
        JSONObject object = new JSONObject();

        for (Map.Entry<String, MessageStatistic> entry : statistics.entrySet()) {
            try {
                JSONArray array = new JSONArray();

                entry.getValue().messages().forEach((m) -> {
                    try {
                        JSONObject obj = new JSONObject();

                        obj.put("contents", m.contents());
                        obj.put("time", m.time());

                        array.put(obj);
                    } catch (JSONException ignored) {
                    }
                });

                object.put(entry.getKey(), array);
            } catch (JSONException ex) {
                ex.printStackTrace();
            }
        }

        try {
            Files.write(Paths.get("statistics.json"), object.toString().getBytes("UTF-8"));
        } catch (UnsupportedEncodingException ignored) {
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }
}

File: src/main/java/io/mazenmc/skypebot/stat/StatisticsManager.kt
package io.mazenmc.skypebot.stat

import com.samczsun.skype4j.chat.messages.ReceivedMessage
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.concurrent.ConcurrentHashMap

public object StatisticsManager {
    private val statistics: MutableMap<String, MessageStatistic> = ConcurrentHashMap<String, MessageStatistic>()

    public fun logMessage(message: ReceivedMessage) {
        var senderId = message.getSender().getUsername()

        if (!statistics.containsKey(senderId))
            statistics.put(senderId, MessageStatistic(senderId))

        statistics.get(senderId).addMessage(message)
    }

    public fun removeStat(id: String) {
        statistics.remove(id)
    }

    public fun ownerFor(statistic: MessageStatistic): String {
        return statistics.filter { e -> e.getValue().equals(statistic) }.iterator().next().getKey()
    }

    public fun statistics(): Map<String, MessageStatistic> {
        return statistics
    }

    public fun loadStatistics() {
        var source = StringBuilder()

        try {
            for (s in Files.readAllLines(Paths.get("statistics.json"))) {
                source.append(s).append("\n")
            }
        } catch (ex: IOException) {
            ex.printStackTrace()
        }

        try {
            var obj = JSONObject(source.toString())
            var keys = obj.keys()

            while (keys.hasNext()) {
                var ob = keys.next()

                if (ob !is String)
                    continue

                statistics.put(ob, MessageStatistic(ob, obj.getJSONArray(ob)))
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    public fun saveStatistics() {
        var json = JSONObject()

        statistics.entrySet().forEach { e ->
            var array = JSONArray()

            e.getValue().messages().forEach { m ->
                var obj = JSONObject()

                obj.put("contents", m.contents)
                obj.put("time", m.time)

                array.put(obj)
            }

            json.put(e.getKey(), array)
        }

        Files.write(Paths.get("statistics.json"), json.toString().toByteArray("UTF-8"))
    }
}